doc/monads/readers.lean

```diff
@@ -0,0 +1,196 @@
+  pure <| match (← IO.getEnv name) with
+          | none => ""
+          | some s => s
+
+def loadEnv : IO Environment := do
+  let path ← getEnvDefault "PATH"
+  let home ← getEnvDefault "HOME"
+  let user ← getEnvDefault "USER"
+  pure { path, home, user }
+
+def func1 (e : Environment) : Float :=
+  let l1 := e.path.length
+  let l2 := e.home.length * 2
+  let l3 := e.user.length * 3
+  (l1 + l2 + l3).toFloat * 2.1
+
+def func2 (env : Environment) : Nat :=
+  2 + (func1 env).floor.toUInt32.toNat
+
+def func3 (env : Environment) : String :=
+  "Result: " ++ (toString (func2 env))
+
+def main : IO Unit := do
+  let env ← loadEnv
+  let str := func3 env
+  IO.println str
+
+#eval main -- Result: 7538
+
+/-!
+The only function actually using the environment is func1. However func1 is a pure function. This
+means it cannot directly call loadEnv, an impure function in the IO monad. This means the
+environment has to be passed through as a variable to the other functions, just so they can
+ultimately pass it to func1. In a language with global variables, you could save env as a global
+value in main. Then func1 could access it directly. There would be no need to have it as a parameter
+to func1, func2 and func3. In larger programs, these "pass-through" variables can cause a lot of
+headaches.
+
+## The Reader Solution
+
+The `ReaderM` monad solves this problem. It effectively creates a global read-only value of a
+specified type. All functions within the monad can "read" the type. Let's look at how the `ReaderM`
+monad changes the shape of this code. Now the functions **no longer need** to be given the
+`Environment` as an explicit parameter, as they can access it through the monad.
+-/
+
+def readerFunc1 : ReaderM Environment Float := do
+  let env ← read
+  let l1 := env.path.length
+  let l2 := env.home.length * 2
+  let l3 := env.user.length * 3
+  return (l1 + l2 + l3).toFloat * 2.1
+
+def readerFunc2 : ReaderM Environment Nat :=
+  readerFunc1 >>= (fun x => return 2 + (x.floor.toUInt32.toNat))
+
+def readerFunc3 : ReaderM Environment String := do
+  let x ← readerFunc2
+  return "Result: " ++ (toString x)
+
+def main2 : IO Unit := do
+  let env ← loadEnv
+  let str := readerFunc3 env
+  IO.println str
+
+#eval main2 -- Result: 7538
+/-!
+`main2` loads the environment as before, and estabilishes the `ReaderM` state by passing `env` to
+readerFunc3.  The `do` notation you learned about in [Monads](monads.lean.md) is hiding some things
+here.  Technically `readerFunc3` is a monadic action and in order to "run" that action the `ReaderM`
+monad provides a `run` method and it is the `ReaderM` run method that takes the initial
+`Environment` state.  So you can actually write the complete code by doing this: `let str :=
+readerFunc3.run env`.
+
+> **Side note 1**: The `return` statement used above also needs some explanation.  The `return`
+statement in Lean can only be used inside a `do` block and is closely related to `pure`, but a
```

Comment:
This isn't true, `return` alone acts as if it had an implicit `do` block around it (i.e. the only difference to `pure` is the parsing precedence). Though I understand that you might not want to lead with this extensions. Perhaps you could remove the "only" and mention return-without-do at the end of the paragraph or elsewhere.

Reply:
Thanks, I removed the comment about only in `do` blocks.  The rest I got from Henrik who said on the Zulip discussion this:

> Note that unlike in haskell there is actually a semantic difference between return and pure in Lean, first return is a special keyword so I can write `return f x` and it will be parsed as `return (f x)` while `pure f x` gets parsed as `(pure f) x` so you usually have to put in a `pure <| f x`, furthermore return will also cause an early return in a monadic function like in an imperative language while pure will not.